### ISSUE TITLE
Add fast 3D constant padding for NumPy and CPU Tensor volumes

### ISSUE THREAD

**Issue by ternaus** [albumentations-team/albucore#159]
Add a reusable constant-padding primitive for 3D volumes, with native CPU Tensor padding and a NumPy → Torch → NumPy path. Direct `PadIfNeeded3D` measurements show that both can improve performance over the current NumPy implementation.

### Current behavior and proposed scope

AlbumentationsX currently implements `pad_3d_with_params` with [`np.pad`](https://github.com/albumentations-team/AlbumentationsX/blob/2baf8ef5bdb44e4310a5d5f85e1d43460bae402e/albumentations/augmentations/transforms3d/functional.py). `PadIfNeeded3D` accepts CPU Tensor input through the transform's NumPy fallback: CDHW Tensor → DHWC NumPy view → padding → CDHW Tensor view. Eligible conversions share storage, so this is not inherently two full-volume copies. The fallback also produces non-contiguous CDHW outputs for multi-channel volumes in the measured cases.

Proposed primitive: `pad3d`, following Albucore's existing 3D layout and routing conventions, with explicit `(front, back, top, bottom, left, right)` spatial padding. NumPy volumes use DHWC; CPU Tensor volumes use CDHW. Preserve channel-less DHW input at the AlbumentationsX boundary or explicitly support it in the primitive. Parameter sampling and target synchronization remain in the transform; Compose needs no changes.

For ordinary NumPy input, `torch.from_numpy(volume)` can keep DHWC layout and call `torch.nn.functional.pad` with `(0, 0, left, right, top, bottom, front, back)`, then return `.numpy()`. No channel permutation is needed. Native CDHW Tensor input uses `(left, right, top, bottom, front, back)` directly. The intended consumers include `Pad3D`, `PadIfNeeded3D`, and the existing 3D crop-and-pad helper.

This opportunity was found while investigating [AlbumentationsX #494](https://github.com/albumentations-team/AlbumentationsX/issues/494). Its random-position correctness fix is separate from this kernel proposal.

### One-thread measurements

Environment: Apple M4 Max, 128 GiB RAM, macOS 26.4.1 arm64, Python 3.12.7, Torch 2.13.0, NumPy 2.5.1, Albucore 0.2.16. Measured September 6, 2026. AlbumentationsX baseline: `2baf8ef5bdb44e4310a5d5f85e1d43460bae402e` with pending padding-correctness fixes. The centered, minimum-size-only examples below are unaffected by those fixes. Albucore source inspected at `d65d2374125b4ca90cf3dc7c1caf0b6a00be268d`.

Both padding backends run with **one CPU compute thread**: Torch reports intra-op=1 and inter-op=1, and OMP/BLAS thread limits are set to 1. OpenCV is not used by these padding paths; this macOS GCD build reports 16 threads even after `cv2.setNumThreads(1)`, so that setting is not evidence of an effective OpenCV thread limit. Calls are directly to `PadIfNeeded3D(volume=...)`, without Compose. Sampling, dispatch, conversions, allocation, padding, and output conversion are included; input creation, transform construction, and correctness comparisons are excluded. Five alternating paired rounds follow warmup and calibration. Ratios below divide the baseline median by the candidate median.

| Input | Cases | Torch faster by >5% | Within 5% | Median ratio across cases |
|---|---:|---:|---:|---:|
| CPU Tensor CDHW: NumPy fallback vs native Torch | 40 | 40 | 0 | 2.32× |
| NumPy DHWC: `np.pad` vs NumPy → Torch → NumPy | 40 | 37 | 3 | 1.44× |
| NumPy DHW controls | 3 | 3 | 0 | 1.14× |

Each 40-case matrix includes D/H/W sizes `(16,24,32)`, `(64,96,128)`, and `(128,192,256)`; C=1/3/5; uint8/float32; contiguous and positive-stride inputs; plus three random-position controls and one no-padding control. Main cases add D=8, H=16, W=24 voxels. Scalar fill is 7 for uint8 and 0.25 for float32. Exact output values, shapes, and dtypes match in all 83 pairs, and inputs remain unchanged. This probe covers volume inputs and scalar fills; masks and other fill forms still require compatibility validation.

Representative centered, contiguous float32 cases; median [min–max] milliseconds across five rounds:

| Input and shape | Current NumPy route | Torch candidate | Ratio |
|---|---:|---:|---:|
| Tensor `(1,16,24,32)` | 0.056 [0.055–0.065] | 0.023 [0.022–0.026] | 2.47× |
| Tensor `(3,64,96,128)` | 2.235 [2.042–2.319] | 0.389 [0.373–0.407] | 5.75× |
| Tensor `(5,128,192,256)` | 28.243 [27.550–30.621] | 14.634 [14.079–21.675] | 1.93× |
| NumPy `(16,24,32,1)` | 0.040 [0.040–0.041] | 0.021 [0.021–0.021] | 1.94× |
| NumPy `(64,96,128,3)` | 0.452 [0.438–0.455] | 0.365 [0.354–0.413] | 1.24× |
| NumPy `(128,192,256,5)` | 17.608 [16.878–17.785] | 17.503 [16.836–18.187] | 1.01× |

Large contiguous NumPy arrays can approach parity: uint8 `(128,192,256,3)` measured 2.433 ms for NumPy versus 2.496 ms for Torch, a 0.97× ratio within the 5% band. These are measurements on one arm64 machine. Repeat paired one-thread measurements on Linux x86 before making a general routing claim. Downstream model consumption was not timed.

### Compatibility and acceptance criteria

- Preserve container, dtype, spatial values, input ownership, and channel axis. Return the input immediately when every padding value is zero.
- Support existing volume dtypes and define mask dtype coverage explicitly, including int16 where needed by AlbumentationsX. Avoid whole-volume dtype conversions for a fill-value problem.
- Preserve existing scalar-fill behavior or explicitly resolve a contract change before integration. For uint8 input, the current AX/NumPy path maps fill `-1` to 255 and `256` to 0; direct Torch padding rejects both as overflow. Both truncate fill `7.5` to 7 in the tested environment.
- Resolve tuple-fill semantics explicitly. The current NumPy call interprets `(3, 7)` as before/after fill values, not per-channel values. A scalar-only Torch replacement must not silently reinterpret accepted inputs.
- Cover positive and negative NumPy strides, read-only arrays, non-contiguous Tensor inputs, and output lifetime. Keep the common supported path free of unnecessary copies; use a compatible path where direct NumPy-to-Torch wrapping is unsuitable.
- Verify outputs remain usable as training inputs; avoid creating inference-only tensors inside the primitive.
- Add focused parity tests for the supported contract and paired one-thread benchmarks that include the complete NumPy bridge. Keep random sampling and Compose outside this kernel change.

### Reproducing the two candidate paths

The following standalone probe runs one centered float32 case for each input container. It uses the same candidate handlers and paired timing method as the matrix. Run it with AlbumentationsX from the baseline revision above and the listed dependencies. Change `shape` and `channels` to reproduce the representative cases. The type annotations on the handlers let the transform use the intended native Tensor or NumPy route.

<details>
<summary>Direct-transform benchmark, one CPU thread</summary>

```python
import gc
import math
import os
import statistics
import time

for variable in (
    "OMP_NUM_THREADS",
    "OPENBLAS_NUM_THREADS",
    "MKL_NUM_THREADS",
    "VECLIB_MAXIMUM_THREADS",
    "NUMEXPR_NUM_THREADS",
):
    os.environ[variable] = "1"

import albumentations as A
import numpy as np
import torch
import torch.nn.functional as F

torch.set_num_threads(1)
torch.set_num_interop_threads(1)
assert torch.get_num_threads() == torch.get_num_interop_threads() == 1
print("Torch intra-op/inter-op threads:", torch.get_num_threads(), torch.get_num_interop_threads())


class TensorPad(A.PadIfNeeded3D):
    def apply_to_volume(
        self,
        volume: torch.Tensor,
        padding: tuple[int, int, int, int, int, int],
        **params: object,
    ) -> torch.Tensor:
        if padding == (0, 0, 0, 0, 0, 0):
            return volume
        front, back, top, bottom, left, right = padding
        return F.pad(volume, (left, right, top, bottom, front, back), value=float(self.fill))


class NumpyPad(A.PadIfNeeded3D):
    def apply_to_volume(
        self,
        volume: np.ndarray,
        padding: tuple[int, int, int, int, int, int],
        **params: object,
    ) -> np.ndarray:
        if padding == (0, 0, 0, 0, 0, 0):
            return volume
        front, back, top, bottom, left, right = padding
        torch_padding = (left, right, top, bottom, front, back)
        if volume.ndim == 4:
            torch_padding = (0, 0, *torch_padding)
        return F.pad(torch.from_numpy(volume), torch_padding, value=float(self.fill)).numpy()


def elapsed(transform, source, iterations):
    start = time.perf_counter_ns()
    for _ in range(iterations):
        transform(volume=source)
    return (time.perf_counter_ns() - start) / 1e9


def benchmark(source, candidate, target):
    transforms = [cls(min_zyx=target, position="center", fill=0.25) for cls in (A.PadIfNeeded3D, candidate)]
    is_tensor = isinstance(source, torch.Tensor)
    snapshot = source.clone() if is_tensor else source.copy()
    expected, actual = [transform(volume=source)["volume"] for transform in transforms]
    assert actual.shape == expected.shape
    assert actual.dtype == expected.dtype == source.dtype
    if is_tensor:
        assert torch.equal(actual, expected)
    else:
        np.testing.assert_array_equal(actual, expected)
    del actual, expected
    for transform in transforms:
        elapsed(transform, source, 5)
    slowest = max(elapsed(transform, source, 5) / 5 for transform in transforms)
    iterations = max(100 if slowest < 0.001 else 5, math.ceil(0.2 / slowest))
    samples = [[], []]
    gc.disable()
    try:
        for round_index in range(5):
            for index in ((0, 1) if round_index % 2 == 0 else (1, 0)):
                transforms[index].set_random_seed(137)
                samples[index].append(elapsed(transforms[index], source, iterations) / iterations)
    finally:
        gc.enable()
    if is_tensor:
        assert torch.equal(source, snapshot)
    else:
        np.testing.assert_array_equal(source, snapshot)
    medians = [statistics.median(values) for values in samples]
    print(type(source).__name__, tuple(source.shape), "calls/round:", iterations)
    for name, values in zip(("NumPy baseline", "Torch candidate"), samples, strict=True):
        print(name, "ms:", [round(value * 1e3, 6) for value in values])
    print("baseline/candidate:", medians[0] / medians[1])


shape, channels = (64, 96, 128), 3
target = tuple(size + pad for size, pad in zip(shape, (8, 16, 24), strict=True))
rng = np.random.default_rng(137)
with torch.no_grad():
    benchmark(torch.from_numpy(rng.random((channels, *shape), dtype=np.float32)), TensorPad, target)
    benchmark(rng.random((*shape, channels), dtype=np.float32), NumpyPad, target)
```

</details>

**Comment by ternaus** [comment]
I implemented the primitive and measured complete public paths on the Apple M4 Max described in the issue. The final routing benchmark covers 219 paired cases, including uint8/float32 volumes, int16 masks, tuple fills, zero padding, positive/negative/read-only/Fortran NumPy layouts, and strided Tensor layouts.

`torch.full` plus insertion did not establish a separate winning region over `F.pad`. The selected public routes are: Torch `F.pad` for contiguous CDHW Tensors; a DHWC NumPy allocation/fill/copy route for interleaved multi-channel Tensors; and a measured NumPy-to-Torch bridge only in the NumPy size/layout regions that beat `np.pad`. Tuple fills retain NumPy before/after semantics. The benchmark report records the exact thresholds, raw samples, and the Apple-only limitation; it makes no Linux x86 routing claim.
